keel-web/src/main/kotlin/com/netflix/spinnaker/keel/services/ApplicationService.kt

```diff
@@ -68,93 +69,51 @@ class ApplicationService(
    *
    * This function assumes there's a single delivery config associated with the application.
-   *
-   * The algorithm is as follows:
-   *
-   * For each environment in the delivery config:
-   * |    For each artifact in the environment:
-   * |    |    Build a map of artifact versions by state (e.g. pending, current, previous, etc.).
-   * |    |    For each pair of (state, versions):
-   * |    |    |    For each version in this state:
-   * |    |    |    |    Get and cache an [ArtifactSummaryInEnvironment] for this version in this environment.
-   * |    |    |    |    Create and cache the [ArtifactVersionSummary] for this version.
-   * |    |    Finally, create the [ArtifactSummary] by looking up the "inner" summaries that were built above.
-   * Return a flat list of all the artifact summaries for all environments in the delivery config.
    */
   fun getArtifactSummariesFor(application: String): List<ArtifactSummary> {
     val deliveryConfig = getFirstDeliveryConfigFor(application)
       ?: throw InvalidRequestException("No delivery config found for application $application")
 
     val environmentSummaries = getEnvironmentSummariesFor(application)
 
-    // map of environments to the set of artifact summaries by state
-    val artifactSummariesByEnvironmentAndState = environmentSummaries.associate { environmentSummary ->
-      environmentSummary.name to mapOf(
-        "current" to mutableSetOf<ArtifactSummaryInEnvironment>(),
-        "deploying" to mutableSetOf<ArtifactSummaryInEnvironment>(),
-        "pending" to mutableSetOf<ArtifactSummaryInEnvironment>(),
-        "approved" to mutableSetOf<ArtifactSummaryInEnvironment>(),
-        "previous" to mutableSetOf<ArtifactSummaryInEnvironment>(),
-        "vetoed" to mutableSetOf<ArtifactSummaryInEnvironment>()
-      )
-    }
-
-    val versionSummariesByArtifact = environmentSummaries.flatMap { environmentSummary ->
-      environmentSummary.artifacts.map { artifact ->
-        artifact.key to mutableSetOf<ArtifactVersionSummary>()
-      }
-    }.toMap()
-
-    val artifactSummaries = environmentSummaries.flatMap { environmentSummary ->
-
-      environmentSummary.artifacts.map { artifact ->
-        val versionsByState = mapOf(
-          "current" to artifact.versions.current?.let { listOf(it) }.orEmpty(),
-          "deploying" to artifact.versions.deploying?.let { listOf(it) }.orEmpty(),
-          "pending" to artifact.versions.pending,
-          "approved" to artifact.versions.approved,
-          "previous" to artifact.versions.previous,
-          "vetoed" to artifact.versions.vetoed
-        )
-
-        versionsByState.entries.map {
-          val state = it.key
-          val versions = it.value
+    return deliveryConfig.artifacts.map { artifact ->
+      artifactRepository.versions(artifact).map { version ->
+        val artifactSummariesInEnvironments = mutableSetOf<ArtifactSummaryInEnvironment>()
```

Comment:
Hopefully this is now much simpler to follow as it's just looping through the entities in the order they appear in the result.

Reply:
Certainly seems a ton easier